### PR TITLE
feat: add button to hide single jobs

### PR DESCRIPTION
**Pull request by sweiler**
Hi there,

we're using gitactionboard in our team (in the otto payments space) as well and have the issue that we just have some pipelines that are supposed to stay red for now. To not let those distract you all the time, it would be great to have a way to hide jobs that you are not interested in.
This PR is adding this, implemented via the preferences-store in the frontend, so that every user can decide for themselves which jobs to hide and which to show.

Is this a feature you would be happy to add, by merging this?
I'm also very open to feedback on this, if you think it would make sense, but should be done differently.

### CODE REVIEW

**Comment by sumanmaity1234** [comment]
Hi @sweiler, can you please add couple of screenshots

**Comment by sweiler** [comment]
Sure, here you go:
Bildschirmfoto 2022-08-23 um 09 31 08

Bildschirmfoto 2022-08-23 um 09 32 38

**Comment by sumanmaity1234** [comment]
Hi @sweiler, Thanks for the screenshots. I have a small questions, what will happen if there is no hidden jobs? Will it still display the 2nd section like **0 hidden jobs**?

**Comment by sweiler** [comment]
No, if you have not hidden any jobs, the section including the horizontal line above it will not be shown.

**Comment by sweiler** [comment]
So @sumanmaity1234 , what do you think would it make sense to include this feature like this?

**Comment by sumanmaity1234** [comment]
Hi @sweiler, this is an interesting feature and I think it'll be a nice addition to current dashboard.

I'll merge this PR this week once I get some time. Meanwhile I'll review the code and if required I'll comment. I hope that'll be fine with you.

File: frontend/src/components/Jobs.vue
Comment (sumanmaity1234) [inline frontend/src/components/Jobs.vue:26]: I am not sure should we call it **hidden jobs** or not

File: frontend/src/components/Jobs.vue
Comment (sumanmaity1234) [inline frontend/src/components/Jobs.vue:47]: We can use **computed** property

File: frontend/src/components/Jobs.vue
Comment (sumanmaity1234) [inline frontend/src/components/Jobs.vue:142]: We can have **jobs** as a state variable and then we can use **computed** to figure out **visible** and **hidden** jobs

File: frontend/src/components/Jobs.vue
Comment (sumanmaity1234) [inline frontend/src/components/Jobs.vue:167]: Instead of having **margin-bottom** for jobs, we should use **margin-top** for hidden jobs

File: frontend/src/components/Jobs.vue
Comment (sumanmaity1234) [inline frontend/src/components/Jobs.vue:172]: Do we really need this styling for hidden jobs? as hidden jobs are also jobs. According to me it should have the same style as normal jobs

File: frontend/src/services/preferences.js
Comment (sumanmaity1234) [inline frontend/src/services/preferences.js:39]: we can use **const** instead of let.

Also can we pull _JSON.parse(this.__get__("hidden-jobs") || "[]");_ into a another method

**Review by sumanmaity1234 — CHANGES_REQUESTED** [review]
Hi @sweiler, I am yet to check the css code base and quickly I checked the JS code and add comments according to my understanding.
Please let me know if you disagree/need more info about certain comment(s)

File: frontend/src/components/Jobs.vue
Comment (sweiler) [reply frontend/src/components/Jobs.vue]: Changed accordingly

File: frontend/src/components/Jobs.vue
Comment (sweiler) [reply frontend/src/components/Jobs.vue]: Changed accordingly

File: frontend/src/components/Jobs.vue
Comment (sweiler) [reply frontend/src/components/Jobs.vue]: Changed accordingly

File: frontend/src/services/preferences.js
Comment (sweiler) [reply frontend/src/services/preferences.js]: Changed accordingly

File: frontend/src/components/Jobs.vue
Comment (sweiler) [reply frontend/src/components/Jobs.vue:180]: This is the styling for the headline "5 hidden jobs". This does not exist for normal jobs, so thats why we have this styling here.

**Comment by sweiler** [comment]
Hi @sumanmaity1234, thanks a lot for your review. I've adapted most of the code accordingly and added two comments to your remarks. Please have a look whenever you are free. If you'd like we could also have a quick chat later today via Otto Teams.

File: frontend/src/components/Jobs.vue
Comment (sumanmaity1234) [reply frontend/src/components/Jobs.vue:180]: ooh my bad, I thought it was for all the fonts (not just header). I'll mark this comment as resolve

File: frontend/src/components/Jobs.vue
Comment (sweiler) [reply frontend/src/components/Jobs.vue]: @sumanmaity1234 I think this is the only open point now, what do you think?

Comment (sumanmaity1234) [reply frontend/src/components/Jobs.vue]: Hi @sweiler , I will think little bit more about it also talk with couple of other contributor, if everyone thinks fine then I'll mark this comment as resolve

**Review by sumanmaity1234 — APPROVED** [review]

**Comment by sweiler** [comment]
Hi @sumanmaity1234 , great to see you approved the PR :)
Is there anything I need to do before it is able to be merged and released?

**Comment by sumanmaity1234** [comment]
Hi @sweiler, I checked the code, I tried to run in local, there is a small issue (I believe) in terms of page refresh. 
It takes sometime once I click on the show/hide button, it takes time to hide/show it.

I'll ping you on teams so that we can connect and discuss about it

**Comment by sweiler** [comment]
@sumanmaity1234 Please have a look at my recent changes. Now the responsibility of hiding this is in the generic Dashboard component, but just disabled for the secrets part. Its now working instantaneously, and gets only synced to the preferences in the background.
I'm not sure if it makes sense this way, as now the Dashboard has this hiding feature, although only one of the use-cases needs it. There is a property 'hide-by-key' which is used to enable the feature and specify which key of the underlying content object is used for hiding.

File: frontend/src/components/Dashboard.vue
Comment (sumanmaity1234) [inline frontend/src/components/Dashboard.vue:25]: Dashboard is a generic component, personally I'll prefer not to hard code like **hidden jobs**

File: frontend/src/components/Job.vue
Comment (sumanmaity1234) [inline frontend/src/components/Job.vue:138]: svg styles can be moved to **HideIcon** component

File: frontend/src/components/Dashboard.vue
Comment (sumanmaity1234) [inline frontend/src/components/Dashboard.vue:9]: Please use plural from for list, ex: **visibleContents**

File: frontend/src/components/Dashboard.vue
Comment (sumanmaity1234) [inline frontend/src/components/Dashboard.vue:21]: Please use plural from for list, ex: **hiddenContents**

File: frontend/src/components/Dashboard.vue
Comment (sumanmaity1234) [inline frontend/src/components/Dashboard.vue:34]: Please use plural from for list, ex: **hiddenContents**

File: frontend/src/components/Job.vue
Comment (sumanmaity1234) [inline frontend/src/components/Job.vue:74]: Any specific reason to use **flex-grow**?

**Review by sumanmaity1234 — CHANGES_REQUESTED** [review]
Hi @sweiler, I added couple of comments which can be improved.

About **Dashboard** generic component I am not sure is it good to have code which is not common across other dashboard.

I am thinking loud does it make sense to make this feature configurable using the client preference page? By default it can be enabled, if someone doesn't like it they can easily disable it from there preference page.

File: frontend/src/components/Dashboard.vue
Comment (sweiler) [reply frontend/src/components/Dashboard.vue]: Solved

File: frontend/src/components/Job.vue
Comment (sweiler) [reply frontend/src/components/Job.vue]: Solved. I also combined HideIcon and ShowIcon into one component, so that they share this style.

File: frontend/src/components/Dashboard.vue
Comment (sweiler) [reply frontend/src/components/Dashboard.vue]: Done

File: frontend/src/components/Dashboard.vue
Comment (sweiler) [reply frontend/src/components/Dashboard.vue]: Done

File: frontend/src/components/Job.vue
Comment (sweiler) [reply frontend/src/components/Job.vue:72]: This is to style the jobs in a way that the hide button is always at the bottom right corner. Without this, the hide button would be at a different position for each job, depending on how many lines the job name takes.

**Comment by sweiler** [comment]
To the "hide-feature in generic Dashboard component" thing: I see only two options here:
- As it is now: Dashboard offers generic feature to hide items, but this is only used by one of the two Dashboards
- Do not have a generic Dashboard component
Maybe you see another way of extracting the hide logic without having two completely separate dashboard implementations.

To the question of making this optional. I think the question here is does those hide buttons annoy people that do not want to use them. If so, can we maybe style them in a less distracting way? Because I think the feature itself is already optional by design. If you don't click any of those buttons, nothing will change, right?

File: frontend/src/components/Dashboard.vue
Comment (sumanmaity1234) [inline frontend/src/components/Dashboard.vue:166]: is it possible to make it generics? **hiddenJobs** is related to workflows

File: frontend/src/icons/ShowIcon.vue
Comment (sumanmaity1234) [inline frontend/src/icons/ShowIcon.vue:22]: Given that there is **HideOrShowIcon** component, **ShowIcon** is it getting used any where?

**Review by sumanmaity1234 — CHANGES_REQUESTED** [review]

File: frontend/src/components/Dashboard.vue
Comment (sweiler) [reply frontend/src/components/Dashboard.vue]: Solved. There is now a hiddenElements object, which currently only holds an entry 'jobs' which are all the hidden jobs. If another dashboard would use this feature, it would store its hidden elements in a different key.

File: frontend/src/icons/ShowIcon.vue
Comment (sweiler) [reply frontend/src/icons/ShowIcon.vue]: True, deleted it.

**Review by sumanmaity1234 — APPROVED** [review]
Changes are look good to me, will test it locally sometime this week and then merge the PR

**Comment by sweiler** [comment]
Great :) Looking forward to show this to the team soon :D

**Comment by sweiler** [comment]
@sumanmaity1234 Fixed the hide button formatting issue now, please have a look :)

**Comment by sumanmaity1234** [comment]
@sweiler Changes looks good to me. Will merge this PR this week

**Review by sumanmaity1234 — APPROVED** [review]

**Pull request merged**
